LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/ConcentratorDeduplication.cs

```diff
@@ -0,0 +1,71 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace LoRaWan.NetworkServer
+{
+    using LoRaWan.NetworkServer.BasicsStation;
+    using Microsoft.Extensions.Caching.Memory;
+    using Microsoft.Extensions.Logging;
+    using System;
+    using System.Security.Cryptography;
+    using System.Text;
+
+    public sealed class ConcentratorDeduplication : IConcentratorDeduplication, IDisposable
+    {
+        internal readonly MemoryCache Cache;
+
+        private readonly ILogger<IConcentratorDeduplication> Logger;
+        private readonly TimeSpan CacheEntryExpiration;
+
+        public ConcentratorDeduplication(
+            ILogger<IConcentratorDeduplication> logger,
+            int cacheEntryExpirationInMilliSeconds = 60_000)
```

Comment:
Agree about the order of logger. 

About `TimeSpan` - it can't be used as an optional parameter because there is no compile time constant (afaik). I need to provide this parameter only from the tests in order to test the expiration policy (in a reasonable time).

Reply:
The traditional way to deal with this is to use constructor overloading and chaining:

```c#
public ConcentratorDeduplication(ILogger<IConcentratorDeduplication> logger) :
    this(logger, TimeSpan.FromMinutes(1)) { }

public ConcentratorDeduplication(ILogger<IConcentratorDeduplication> logger,
                                 TimeSpan cacheEntryExpiration)
{
    this.Cache = new MemoryCache(new MemoryCacheOptions());
    this.Logger = logger;
    this.CacheEntryExpiration = cacheEntryExpiration;
}
```

Another option, if you wish to stick with optional arguments, is to make `TimeSpan` nullable and then use `null` as the default value. In the implementation body of the constructor, you can then default a specific value in case of `null`. Below is an example of what I mean:

```c#
public ConcentratorDeduplication(ILogger<IConcentratorDeduplication> logger,
                                 TimeSpan? cacheEntryExpiration = null)
{
    this.Cache = new MemoryCache(new MemoryCacheOptions());
    this.Logger = logger;
    this.CacheEntryExpiration = cacheEntryExpiration ?? TimeSpan.FromMinutes(1);
}
```

Reply:
Thanks very much for the snippets - I have seen these patterns but it hadn't clicked why we need those. Went with the second variant in df55a3bcea66a1a83f2d54939f71415da3595a7a